vispy/glsl/arrowheads/arrowheads.vert

```diff
@@ -75,10 +75,9 @@ void main (void)
     v_color       = color;
     v_linewidth   = linewidth;
 
-    vec2 body = v2 - v1;
+    vec3 body = $transform(vec4(v2 - v1, 0.)).xyz;
```

Comment:
* The `transform` is a matrix product and the multiplication is distributive so `transform * (v2 - v1) = transform * v2 - transform * v1`
* From a geometry perspective, if you apply the transformation to the v2 and v1 points, it should be equivalent to apply the transformation to the vector defined by v2 - v1
* A small script to test with the transformation of a turntable and panzoom camera :
 ```python
import numpy as np
from vispy import scene

v2 = np.random.rand(4)
v1 = np.random.rand(4)

cam_pz = scene.cameras.PanZoomCamera()
cam_tt = scene.cameras.TurntableCamera()

for k in [cam_pz, cam_tt]:
    tr = k.transform.map(v2) - k.transform.map(v1)
    tr_diff = k.transform.map(v2 - v1)
    assert np.array_equal(tr, tr_diff)
```
* If this is really a problem for you, I'm not against `$transform(v2) - $transform(v1)` but, from my understanding, apply this transformation twice is not a necessity.

Reply:
About the tests, I tried on my machine to run `test_arrows.py` but `assert_image_approved` failed with the following error :
`TypeError: Test result shape [ 1033.  1916.     4.] is not an integer factor larger than standard image shape [ 100.  100.    4.].`

I also tried to run with `--vispy-audit` :

![vispy canvas_005](https://user-images.githubusercontent.com/15892073/32174054-181838f6-bd81-11e7-87c1-ebb23a577cb7.png)

* Line color is ok
* Arrow color is not ok because I didn't turn the `arrow_color` input to 'red'
* The main issue is the canvas size difference between the render and this is related to my system. I use [i3](https://i3wm.org/), which is a tilling window manager and by default it maximized the window. So I switch to the linux Cinnamon desktop and all tests finally passed

Reply:
Everything you're saying makes sense in the cases you are testing with. There are however transforms that are not matrix products or are not linear. The one that comes to mind is the MagnifyCamera. Although not used often, I don't think your code would work with that transform.

There are also transforms where certain areas of the canvas are not valid coordinates. For example a geographic projection transformation from lon/lat space to a certain map projection. A lon/lat of `(-95, 45)` may be valid in the projection space, but if this is changed to a distance vector with another point then you might get `(-0.05, 0.234)` in lon/lat space, but this is not a valid coordinate in the projection.

Reply:
Ah ok, right, in that case I can write `$transform(v2) - $transform(v1)`.

Reply:
See 9cb66cb76494d1b284807ac1aebe73b1f38c6eb5